util.sh

```diff
@@ -0,0 +1,147 @@
+# without needing to build+tag images
+function compose {
+    if [ -n "${CI:-}" ]; then
+        # sed uncomments lines starting with "# CI"
+        # and removes lines ending with "# CI"
+        # therefore allowing to switch what is used locally and in CI
+        set -x
+        exec docker-compose \
+            -f <(cat docker-compose.yml | sed -r -e 's/^(\s+)# CI\s*(.*)/\1\2/g' -e '/# CI/d') \
+            --project-directory $PWD \
+            $@
+    else
+        set -x
+        exec docker-compose $@
+    fi
+}
+
+# if Makefile target specifies which docker-compose service to use
+# get its name to be used with docker-compose
+# otherwise fallback to provided default
+# usage:
+# service_for_compose <default> "$@"
+function service_for_compose {
+    default=$1
+    args=$2
+    # looks magical but its pretty simple
+    # for example if script is called with ./build.sh lint
+    # it will look for a line in Makefile starting with "lint: # docker-compose:"
+    # and if it finds such a line, it extracts docker-compose service
+    # which is expected to be used for that makefile target
+    service=$(
+        (
+            grep -E "^($(
+                echo $args \
+                    | tr ' ' '\n' \
+                    | paste -s -d'|'
+            )): # docker-compose:" Makefile \
+                || true
+        ) \
+            | cut -d: -f3 \
+            | head -n1
+    )
+    echo ${service:-$default}
+}
+
+# ============================================================================
+# HELP
+# ============================================================================
+
+ORANGE="\033[0;33m"
+BLUE="\033[0;34m"
+RED="\033[0;31m"
+END_COLOR="\033[0m"
+if [ -n "${NO_COLOR:-}" ]; then
+    ORANGE=
+    BLUE=
+    RED=
+    END_COLOR=
+fi
+
+HELP_WIDTH=${HELP_WIDTH:-15}
+
+function _help_format {
+    cat - | awk "{printf \"${BLUE}%-${HELP_WIDTH}s${END_COLOR} %s\n\", \$1, substr(\$0, length(\$1) + 1);}"
+}
+
+function _help_makefile {
+    if [ ! -f Makefile ]; then
+        return
+    fi
+    grep -E '^[a-zA-Z_-]+:.*?## .*$$' Makefile* \
+        | cut -d: -f2- \
+        | sort \
+        | sed 's/:\s*##\s*/ /g' \
+        | _help_format
+
+}
+
+function _help_commands {
+    bin=$0
+    grep -E '^\s*## help:command' $bin \
+        | cut -d: -f3- \
+        | sort \
+        | _help_format \
+        || true
+}
+
+function _help_flags {
+    bin=$0
+    grep -E '^\s*## help:flag' $bin \
+        | cut -d: -f3- \
+        | sort \
+        | _help_format \
+        || true
+}
+
+# combine help from multiple sources
+function _help_combined {
+    echo -e $ORANGE$(basename $PWD)$END_COLOR
+    echo
+    echo -e $ORANGE$0$END_COLOR [flag ...] [command ...]
+    echo
+
+    echo -e ${ORANGE}Flags:${END_COLOR}
+    echo
+    {
+        echo -h/--help print this message | _help_format
+        _help_flags $@
+    } | sort
+    echo
+
+    echo -e ${ORANGE}Commands:$END_COLOR
```

Comment:
Can you make `$var` consistent either with or without the braces? My personal preference is for with the braces, especially in the cases here where inserting the colors is harder to read without the braces, but I'm fine with whatever as long as it's consistent. Not just in this function, but throughout.

Reply:
adjusted. just left `$@` and couple of other standalong places without braces but all strings I think adjusted to use `${}`